app/models/questionnaire.rb

```diff
@@ -184,4 +187,28 @@ def all_scores_have_known_operations
 
     errors.add(:content, "the following scores have an unknown operation: #{result.pretty_inspect}")
   end
+
+  def all_scores_have_valid_ids_in_preprocessing
+    result = content[:scores]
+             .select { |score| score.key?(:preprocessing) }
+             .select { |score| (score[:preprocessing].keys - (score[:ids] || [])).present? }
+             .map { |score| score[:label] || score[:id] }
+    return if result.blank?
+
+    errors.add(:content, "the following scores have invalid ids in preprocessing steps: #{result.pretty_inspect}")
+  end
+
+  def all_scores_have_valid_preprocessing
+    result = content[:scores]
+             .select { |score| score.key?(:preprocessing) }
+             .select { |score| invalid_preprocessing_steps?(score[:preprocessing]) }
+             .map { |score| score[:label] || score[:id] }
+    return if result.blank?
+
+    errors.add(:content, "the following scores have invalid preprocessing steps: #{result.pretty_inspect}")
+  end
+
+  def invalid_preprocessing_steps?(score_preprocessing)
```

Comment:
would it not make sense to change this to something like `only_valid_preprocessing_steps?`?

Reply:
And then change it to .reject instead of .select, right? Sure, done.